### PR TITLE
[Diagnostics] Display correct debug note for compound references typo suggestions

### PR DESCRIPTION
**Pull request by OscarSwanros**
This PR tries to address the issue where functions with multiple arguments would just be diagnosed incorrectly ignoring their attributes.

Resolves #45842.

### CODE REVIEW

**Comment by OscarSwanros** [comment]
@swift-ci Please smoke test

File: lib/Sema/TypeCheckNameLookup.cpp [hunk 1]
```diff
@@ -555,13 +555,16 @@ void TypeChecker::performTypoCorrection(DeclContext *DC, DeclRefKind refKind,
 }
 
 static InFlightDiagnostic
-diagnoseTypoCorrection(TypeChecker &tc, DeclNameLoc loc, ValueDecl *decl) {
+diagnoseTypoCorrection(TypeChecker &tc, DeclNameLoc loc, ValueDecl *decl, FunctionRefKind refKind) {
```
Comment (modocache) [inline lib/Sema/TypeCheckNameLookup.cpp]: nit-pick: The style guide for C++ in Swift projects is a 80 character column width. You should format this like so:

```cxx
diagnoseTypoCorrection(TypeChecker &tc,
                       DeclNameLoc loc,
                       ValueDecl *decl,
                       FunctionRefKind refKind) {
```

Same goes for the line of code below this one: you'll need to split it across two lines for it to fit within 80 characters.

**Review by modocache — COMMENTED** [review]
This looks pretty good to me! Just needs a few code style changes. Assuming the tests pass, I think this could be merged -- but I'll leave it to someone more familiar with this part of the codebase to approve the pull request.

**Comment by modocache** [comment]
@swift-ci please smoke test

File: lib/Sema/TypeCheckNameLookup.cpp (hunk 1 above)
Comment (OscarSwanros) [inline lib/Sema/TypeCheckNameLookup.cpp]: Oh, thanks! Will update accordingly.

**Comment by OscarSwanros** [comment]
@modocache updated with code-style comments. Thanks for the eyes on this!

File: lib/Sema/TypeCheckNameLookup.cpp [hunk 2]
```diff
@@ -555,13 +555,21 @@ void TypeChecker::performTypoCorrection(DeclContext *DC, DeclRefKind refKind,
 }
 
 static InFlightDiagnostic
-diagnoseTypoCorrection(TypeChecker &tc, DeclNameLoc loc, ValueDecl *decl) {
+diagnoseTypoCorrection(TypeChecker &tc,
+                       DeclNameLoc loc, 
+                       ValueDecl *decl, 
+                       FunctionRefKind refKind) {
+  
+  auto name = refKind == FunctionRefKind::Compound ? decl->getFullName() :
```
Comment (CodaFi) [inline lib/Sema/TypeCheckNameLookup.cpp]: LLVM style moves the colon to a new line in-line with the else condition and aligns it with the question mark.

File: test/Sema/compound_references.swift [hunk 3]
```diff
@@ -0,0 +1,4 @@
+// RUN: %target-typecheck-verify-swift
+
+func foo(x: Int) {} // expected-note * {{did you mean 'foo(x:)'?}}
+let f = foo(_:) // expected-error {{use of unresolved identifier 'foo'}}
```
Comment (CodaFi) [inline test/Sema/compound_references.swift]: Please remove this file and add these tests to `Sema/typo_correction.swift` (sans * on the `expected-note`).

**Review by CodaFi — CHANGES_REQUESTED** [review]

File: lib/Sema/TypeCheckNameLookup.cpp [hunk 4]
```diff
@@ -555,13 +555,21 @@ void TypeChecker::performTypoCorrection(DeclContext *DC, DeclRefKind refKind,
 }
 
 static InFlightDiagnostic
-diagnoseTypoCorrection(TypeChecker &tc, DeclNameLoc loc, ValueDecl *decl) {
+diagnoseTypoCorrection(TypeChecker &tc,
+                       DeclNameLoc loc,
```
Comment (CodaFi) [inline lib/Sema/TypeCheckNameLookup.cpp]: You don't necessarily have to fold the entire signature over, just those that go over the line-limit.  If it makes it easier, you can run `clang-format` with the default LLVM style to have it wrap these for you.

**Comment by OscarSwanros** [comment]
@CodaFi Fixed your comments. Thanks!

File: test/Sema/typo_correction.swift [hunk 5]
```diff
@@ -84,3 +84,9 @@ _ = [Any]().withUnsafeBufferPointer { (buf) -> [Any] in
   guard let base = buf.baseAddress else { return [] }
   return (base ..< base + buf.count).m // expected-error {{value of type 'CountableRange<_>' has no member 'm'}}
 }
+
+
+// SR-3254: Test for correct diagnostics for compound references
+foo(x: Int) {} // expected-note {{did you mean 'foo(x:)'?}}
```
Comment (CodaFi) [inline test/Sema/typo_correction.swift]: This doesn't test the part of the diagnostic you're interested in.

File: lib/Sema/TypeCheckNameLookup.cpp [hunk 6]
```diff
@@ -555,13 +555,19 @@ void TypeChecker::performTypoCorrection(DeclContext *DC, DeclRefKind refKind,
 }
 
 static InFlightDiagnostic
-diagnoseTypoCorrection(TypeChecker &tc, DeclNameLoc loc, ValueDecl *decl) {
+diagnoseTypoCorrection(TypeChecker &tc, DeclNameLoc loc, ValueDecl *decl,
+                       FunctionRefKind refKind) {
+  
+  auto name = refKind == FunctionRefKind::Compound ? decl->getFullName()
+                                                   : decl->getName();
```
Comment (CodaFi) [inline lib/Sema/TypeCheckNameLookup.cpp]: Here you must add the quotes back or it'll fall any existing typo correction tests that rely on it.

File: lib/Sema/TypeChecker.h [hunk 7]
```diff
@@ -2083,7 +2083,8 @@ class TypeChecker final : public LazyResolver {
                              unsigned maxResults = 4);
 
   void noteTypoCorrection(DeclName name, DeclNameLoc nameLoc,
-                          const LookupResult::Result &suggestion);
+                          const LookupResult::Result &suggestion,
+                          FunctionRefKind refKind = FunctionRefKind::Unapplied);
```
Comment (CodaFi) [inline lib/Sema/TypeChecker.h]: I'm still not 100% sure this is the right type for this.  For names that aren't function references that go through typo correction this is kind of nonsensical.

**Review by CodaFi — CHANGES_REQUESTED** [review]

File: lib/Sema/TypeChecker.h (hunk 7 above)
Comment (OscarSwanros) [inline lib/Sema/TypeChecker.h]: Is there any other type that would work in this case?

**Comment by CodaFi** [comment]
@OscarSwanros What's the status of this pull request?  If you're still interested in this bug fix, could you rebase this pull request and give me a ping?

**Comment by CodaFi** [comment]
This patch has fallen behind after a month.  @OscarSwanros feel free to reopen this when you find the time.

**Comment by OscarSwanros** [comment]
Oh, sorry @CodaFi — have been swamped and missed your notification. Will try to find the time soon.

**Pull request closed without merging**